src/js/backup_restore.js

```diff
@@ -321,25 +321,24 @@ function configuration_restore(callback) {
                 if ((e.total != 0 && e.total == e.loaded) || GUI.isCordova()) {
                     // Cordova: Ignore verification : seem to have a bug with progressEvent returned
                     console.log('Read SUCCESSFUL');
-
+                    let configuration;
                     try { // check if string provided is a valid JSON
-                        var configuration = JSON.parse(e.target.result);
-                    } catch (e) {
+                        configuration = JSON.parse(e.target.result);
+                    } catch (err) {
                         // data provided != valid json object
-                        console.log('Data provided != valid JSON string, restore aborted.');
+                        console.log(`Data provided != valid JSON string, restore aborted: ${err}`);
 
                         return;
                     }
 
-
                     // validate
-                    if (typeof configuration.generatedBy !== 'undefined' && compareVersions(configuration.generatedBy, CONFIGURATOR.BACKUP_FILE_VERSION_MIN_SUPPORTED)) {
+                    if (typeof configuration.generatedBy && compareVersions(configuration.generatedBy, CONFIGURATOR.BACKUP_FILE_VERSION_MIN_SUPPORTED)) {
```

Comment:
I'm not a JS expert
but I think this is not correct.
`typeof configuration.generatedBy` is always true because typeof always returns non-empty string even if the variable is undefined.

Example:
```
let a = undefined;

if (typeof a) {
	console.log("(typeof a) is true");
}

if (typeof a !== 'undefined') {
	console.log("(typeof a !== 'undefined) is true");
}

```

output:
`(typeof a) is true`


`typeof configuration.generatedBy !== 'undefined'`
is the "old" way to check if variable is undefined.
I think now it can be replaced with
`configuration.generatedBy !== undefined`

Reply:
One reason to still use `typeof` is that it does not throw an error if the variable has not been declared.
```
let a = undefined;
// x is not declared

if (a === undefined) {
	console.log(`a is undefined`);
}
if (typeof x === `undefined`) {
	console.log(`x is undefined`);
}
if (typeof x) {
	console.log('x is still undefined');
}
if (x) {
        console.log('this will throw an error');
}
```
Output:
```
"a is undefined"
"x is undefined"
"x is still undefined"
Uncaught ReferenceError: x is not defined"
```

Reply:
But there is a strange difference between `undefined` and `'undefined'`

Reply:
the problem is here:
```
if (typeof x) {
	console.log('x is still undefined');
}
```
this one will ALWAYS be true no matter what you do with X

Reply:
I am reading the book Understanding `EcmaScript 6` from Nicholas C. Zakas purchased two years ago.

Reply:
> I am reading the book `Understanding EcmaScript 6` from Nicholas C. Zakas purchased two years ago.

then maybe the only correct way is to leave like it was before?
`typeof configuration.generatedBy !== 'undefined'`

Reply:
Yeah I think you are right. It should not be changed.
(typeof x !== `undefined`)

Reply:
Thanks @limonspb changed after verifying and testing.

Reply:
Ok, gents, I think we need to make something clear here:

```
if (x === undefined) { } // will throw if x has not been defined
```

BUT

```
if (x.y == undefined) {} // will NOT throw if x.y has not been declared (but will throw if x has not been declared)
```

Seeing how ungainly 'typeof' is (string comparison, anybody ever come across `typeof x === 'unedfined'`?), my take is that `typeof` is appropriate for use when checking global variables, the risks of its use outweigh its benefits (i.e. none) when checking properties.